### ISSUE TITLE
Does Superpowered SDK provide methods for resolving audio feedback (Larsen effect)?

### ISSUE THREAD

**Issue by amos-yau** [superpoweredSDK/Low-Latency-Android-Audio-iOS-Audio-Engine#270]
I'm trying the Android sample project "FrequencyDomain", when the speaker turn louder, the feedback is serious, anyway to resolve this?

**Comment by superpoweredSDK** [comment]
Yes, put on your headphones. This example shows how to transform time-domain audio to frequency-domain and vice-versa. It's not intended to be an easy user app.

**Comment by amos-yau** [comment]
Brilliant answer. I'm not going to modify it as an easy user app, I'm working on a video call Android app.

So modify a bit the question: Does Superpowered SDK provide **API call** methods for resolving audio feedback (Larsen effect)?

**Comment by EtherBit** [comment]
@amos-yau you could probably use speedx engine to resolve the audio feedback.

**Comment by amos-yau** [comment]
@EtherBit I've tried but not success to integrate with Superpowered, you know how to use speex engine to resolve the audio feedback?

**Comment by EtherBit** [comment]
I have not used it (yet, but most likely will need it soon), but it is theoretically possible by using the echo cancellation (from the documentation):
#include <speex/speex_echo.h> 

Then, an echo canceller state can be created by:

SpeexEchoState *echo_state = speex_echo_state_init(frame_size, filter_length);

Once the echo canceller state is created, audio can be processed by:

speex_echo_cancellation(echo_state, input_frame, echo_frame, output_frame);

where input_frame is the audio as captured by the microphone, echo_frame is the signal that was played in the speaker (and needs to be removed) and output_frame is the signal with echo removed.

This guy was successful:
https://www.youtube.com/watch?v=e1-HwidHMy0

**Comment by amos-yau** [comment]
@EtherBit Ya I've tried the speex_echo.h, thanks a lot for your post, but it's failed (I'm still finding the reason why...)

Interesting I also found that video, that's why I decided to try this library.

**Comment by EtherBit** [comment]
@amos-yau 
If I ever figure it out (I will need to), I will post it. In the meantime, there seems to be some a lot of repo on Github that use speex. Maybe you could browse through them and use them as basis to figure out what you are doing wrong?

**Comment by amos-yau** [comment]
@EtherBit man you are so kind, appreciate a lot, I will try what you said :)

**Comment by temkich** [comment]
I tried to use speex_echo for echo cancellation in SuperpoweredIOSAudioIO, but I did not succeed. 
And I tried to replace the line
`desc.componentSubType = kAudioUnitSubType_RemoteIO;`
in file SuperpoweredIOSAudioIO.mm, on a new line
`desc.componentSubType = kAudioUnitSubType_VoiceProcessingIO;`
and I got what I wanted. Echo cancellation worked.
But it does not work constantly, once I get an error
`[aurioc] 918: failed: -66635 (enable 3, outf< 2 ch,      0 Hz, Float32, non-inter> inf< 2 ch,      0 Hz, Float32, non-inter>)`
I do not understand what this error is and how can I solve it? 
Why the file SuperpoweredIOSAudioIO code is not used immediately `kAudioUnitSubType_VoiceProcessingIO` ? This does not work for iOS? @superpoweredSDK
